File: modules/chat-core/src/main/java/com/shterneregen/securelan/chat/discovery/DiscoveryCodec.java
package com.shterneregen.securelan.chat.discovery;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Objects;

public final class DiscoveryCodec {
    private static final String PREFIX = "SECURELAN_DISCOVERY_V1";

    private DiscoveryCodec() {
    }

    public static String encode(Message message) {
        Objects.requireNonNull(message, "message");
        return String.join("|",
                PREFIX,
                encodeText(message.peerId()),
                encodeText(message.nickname()),
                Integer.toString(message.chatPort()),
                Integer.toString(message.filePort())
        );
    }

    public static Message decode(String payload) {
        String[] parts = payload == null ? new String[0] : payload.split("\\|", -1);
        if (parts.length != 5 || !PREFIX.equals(parts[0])) {
            throw new IllegalArgumentException("Unsupported discovery message");
        }
        String peerId = decodeText(parts[1]);
        String nickname = decodeText(parts[2]);
        int chatPort = parsePort(parts[3], "chatPort");
        int filePort = parsePort(parts[4], "filePort");
        return new Message(peerId, nickname, chatPort, filePort);
    }

    public static byte[] toBytes(Message message) {
        return encode(message).getBytes(StandardCharsets.UTF_8);
    }

    public static Message fromBytes(byte[] data, int length) {
        Objects.requireNonNull(data, "data");
        if (length < 0 || length > data.length) {
            throw new IllegalArgumentException("length must be between 0 and data length");
        }
        return decode(new String(data, 0, length, StandardCharsets.UTF_8));
    }

    private static String encodeText(String value) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(value.getBytes(StandardCharsets.UTF_8));
    }

    private static String decodeText(String value) {
        return new String(Base64.getUrlDecoder().decode(value), StandardCharsets.UTF_8);
    }

    private static int parsePort(String value, String fieldName) {
        try {
            int port = Integer.parseInt(value);
            if (port < 1 || port > 65535) {
                throw new IllegalArgumentException(fieldName + " must be between 1 and 65535");
            }
            return port;
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(fieldName + " must be a valid TCP/UDP port", e);
        }
    }

    public record Message(String peerId, String nickname, int chatPort, int filePort) {
        public Message {
            Objects.requireNonNull(peerId, "peerId");
            Objects.requireNonNull(nickname, "nickname");
            if (chatPort < 1 || chatPort > 65535) {
                throw new IllegalArgumentException("chatPort must be between 1 and 65535");
            }
            if (filePort < 1 || filePort > 65535) {
                throw new IllegalArgumentException("filePort must be between 1 and 65535");
            }
        }
    }
}

File: modules/chat-core/src/main/kotlin/com/shterneregen/securelan/chat/discovery/DiscoveryCodec.kt
package com.shterneregen.securelan.chat.discovery

import java.nio.charset.StandardCharsets
import java.util.Base64
import java.util.Objects

object DiscoveryCodec {
    private const val PREFIX = "SECURELAN_DISCOVERY_V1"

    @JvmStatic
    fun encode(message: Message): String {
        Objects.requireNonNull(message, "message")
        return listOf(
            PREFIX,
            encodeText(message.peerId),
            encodeText(message.nickname),
            message.chatPort.toString(),
            message.filePort.toString(),
        ).joinToString("|")
    }

    @JvmStatic
    fun decode(payload: String?): Message {
        val parts = payload?.split("|", ignoreCase = false, limit = 0)?.toTypedArray() ?: emptyArray()
        if (parts.size != 5 || PREFIX != parts[0]) {
            throw IllegalArgumentException("Unsupported discovery message")
        }
        val peerId = decodeText(parts[1])
        val nickname = decodeText(parts[2])
        val chatPort = parsePort(parts[3], "chatPort")
        val filePort = parsePort(parts[4], "filePort")
        return Message(peerId, nickname, chatPort, filePort)
    }

    @JvmStatic
    fun toBytes(message: Message): ByteArray = encode(message).toByteArray(StandardCharsets.UTF_8)

    @JvmStatic
    fun fromBytes(data: ByteArray, length: Int): Message {
        Objects.requireNonNull(data, "data")
        require(length in 0..data.size) { "length must be between 0 and data length" }
        return decode(String(data, 0, length, StandardCharsets.UTF_8))
    }

    private fun encodeText(value: String): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(value.toByteArray(StandardCharsets.UTF_8))

    private fun decodeText(value: String): String = String(Base64.getUrlDecoder().decode(value), StandardCharsets.UTF_8)

    private fun parsePort(value: String, fieldName: String): Int = try {
        val port = value.toInt()
        require(port in 1..65_535) { "$fieldName must be between 1 and 65535" }
        port
    } catch (e: NumberFormatException) {
        throw IllegalArgumentException("$fieldName must be a valid TCP/UDP port", e)
    }

    @JvmRecord
    data class Message(
        val peerId: String,
        val nickname: String,
        val chatPort: Int,
        val filePort: Int,
    ) {
        init {
            Objects.requireNonNull(peerId, "peerId")
            Objects.requireNonNull(nickname, "nickname")
            require(chatPort in 1..65_535) { "chatPort must be between 1 and 65535" }
            require(filePort in 1..65_535) { "filePort must be between 1 and 65535" }
        }
    }
}
